addons/account/tests/test_sequence_mixin.py

```diff
@@ -537,3 +554,32 @@ def test_sequence_concurency_no_useless_lock(self):
         # check the values
         moves = env0['account.move'].browse(self.data['move_ids'])
         self.assertEqual(moves.mapped('name'), ['CT/2016/01/0001', '/', 'CT/2016/01/0002'])
+
+    def test_sequence_concurrency_no_deadlock(self):
+        env0, env1, env2 = self.data['envs']
+
+        # start the transactions here on cr1 to simulate concurrency with cr2
+        move = env1['account.move'].browse(self.data['move_ids'][1])
+        move.action_post()
+
+        async def t2():
+            payment = env2['account.payment'].browse(self.data['payment_ids'][1])
+            payment.action_post()
+            move = env2['account.move'].browse(self.data['move_ids'][2])
+            move.action_post()
+            env2.cr.commit()
+        promise = t2()
+
+        asyncio.run(asyncio.sleep(1))  # the first action_post in t2 will never return before this
+
+        payment = env1['account.payment'].browse(self.data['payment_ids'][2])
+        payment.action_post()
+        env1.cr.commit()
+
+        asyncio.run(promise)
```

Comment:
And that seems completely useless, as the coroutine contains literally nothing that's asynchronous (but I don't have that much experience with asyncio, @Julien00859 might know better).

Reply:
The cursors are asynchronous

Reply:
The entire thing is 100% synchronous and does this:

```py
def test_sequence_concurrency_no_deadlock(self):
    env0, env1, env2 = self.data['envs']

    move = env1['account.move'].browse(self.data['move_ids'][1])
    move.action_post()

    time.sleep(1)

    payment = env1['account.payment'].browse(self.data['payment_ids'][2])
    payment.action_post()
    env1.cr.commit()

    payment = env2['account.payment'].browse(self.data['payment_ids'][1])
    payment.action_post()
    move = env2['account.move'].browse(self.data['move_ids'][2])
    move.action_post()
    env2.cr.commit()

    moves = env0['account.move'].browse(self.data['move_ids'])
    self.assertEqual(moves.mapped('name'), ['CT/2016/01/0001', 'CT/2016/01/0002', 'CT/2016/01/0003'])

    moves = env0['account.payment'].browse(self.data['payment_ids'])
    self.assertEqual(moves.mapped('name'), ['PBNK1/2016/00001', 'PBNK1/2016/00003', 'PBNK1/2016/00002'])
```

When you create a coroutine without `await`-ing it, the coroutine is not linked to any running loop. Actually there is no event loop created by default, `asyncio.run` performs the necessary boilerplates to create an event-loop if none was existing. The code `promise = t2()` gives a wrong hint, the variable "promise" merely hold a reference to a coroutine that is not being executed nor scheduled for execution. If you want to schedule it for execution (but not *run it now*) you need to create a task `task = asyncio.create_task(t2())`, the created task would be scheduled for execution and be executed the next time the event loop runs.

The second part `asyncio.run(asyncio.sleep(1))` creates a `sleep` coroutine, start an event loop and run it until the coroutine completes. In the current state, only `asyncio.sleep(1)` runs because the other coroutine `t2()` is not a task of the event-loop. So it runs a single coroutine that sleep 1 second. That's basically `time.sleep(1)`.

Had `t2` be scheduled for execution via `create_task`, it would had been executed *as soon as* the event-loop starts with `run()` with no regard for the `sleep` coroutine. The two coroutines are independents and would have been executed concurrently. That's what asynchronous programming is for. Because `run` would stop once `sleep(1)` finishes, there is no garantee that `t2()` would had been fully executed.

Final note: cursors are not asynchronous. Put simply psycopg2 doesn't support async-await.


Asyncio is not the right tool for the job you're trying to do. If you want to do a concurrency test, please use threads, locks, events and barriers.

Reply:
If you want to have a fine control on the execution flow in order to test again concurrency bugs, I can suggest you use threads, locks, events and barriers. Here is an attempt of my own for a concurrency bug inside the crons https://github.com/odoo/odoo/pull/97087

Reply:
I rewrote it like this
```python
    def test_sequence_concurrency_no_deadlock(self):
        env0, env1, env2 = self.data['envs']

        # start the transactions here on cr1 to simulate concurrency with cr2
        move = env1['account.move'].browse(self.data['move_ids'][1])
        move.action_post()

        def t1():
            payment = env1['account.payment'].browse(self.data['payment_ids'][2])
            payment.action_post()
            env1.cr.commit()

        def t2():
            payment = env2['account.payment'].browse(self.data['payment_ids'][1])
            payment.action_post()
            move = env2['account.move'].browse(self.data['move_ids'][2])
            move.action_post()
            env2.cr.commit()
        thread2 = threading.Thread(target=t2)

        thread2.start()
        time.sleep(1)  # the first action_post in t2 will never return before this
        t1()
        thread2.join()

        moves = env0['account.move'].browse(self.data['move_ids'])
        self.assertEqual(moves.mapped('name'), ['CT/2016/01/0001', 'CT/2016/01/0002', 'CT/2016/01/0003'])

        moves = env0['account.payment'].browse(self.data['payment_ids'])
        self.assertEqual(moves.mapped('name'), ['PBNK1/2016/00001', 'PBNK1/2016/00003', 'PBNK1/2016/00002'])
```
But now I have the deadlock that I wanted to avoid.

Reply:
> But now I have the deadlock that I wanted to avoid.

Unclear what the deadlock is, but cursors are not thread-safe, only connections are: https://www.psycopg.org/docs/usage.html#thread-and-process-safety

Reply:
Cursors are not shared between different threads